File: 98.validate-binary-search-tree.py
#
# @lc app=leetcode id=98 lang=python3
#
# [98] Validate Binary Search Tree
#

# @lc code=start
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def __init__(self):
        self.values = []
        self.call_count = 1
        
    def isValidBST(self, root: TreeNode) -> bool:

        self.call_count += 1
        if not root:
            return True
        if not root.left and not  root.right:
            if self.values and self.values[-1] >= root.val:
                return False
            self.values.append(root.val)
            return True
        if root.left:
            if not self.isValidBST(root.left):
                return False
        if self.values and self.values[-1] >= root.val:
            return False
        self.values.append(root.val)
        if root.right:
            if not self.isValidBST(root.right):
                return False
        return True
        
        
# @lc code=end

# if __name__ == "__main__":
#     right = TreeNode(1)
#     root = TreeNode(1)
#     root.right = right
#     print(Solution().isValidBST(root))

File: golang/98.validate-binary-search-tree.go
/*
 * @lc app=leetcode.cn id=98 lang=golang
 *
 * [98] Validate Binary Search Tree
 */
package main

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isValidBST(root *TreeNode) bool {
	if root == nil {
		return true
	}
	items := []int{}
	ret := true

	collect := func(i int) {
		if len(items) > 0 && items[len(items)-1] >= i {
			ret = false
		}
		items = append(items, i)
	}
	isValidBSTHelper(root, collect)
	return ret
}

func isValidBSTHelper(root *TreeNode, collect func(int)) {
	if root.Left == nil && root.Right == nil {
		collect(root.Val)
		return
	}
	if root.Left != nil {
		isValidBSTHelper(root.Left, collect)
	}

	collect(root.Val)

	if root.Right != nil {
		isValidBSTHelper(root.Right, collect)
	}

}

// @lc code=end
